types/wlr_cursor.c

```diff
@@ -233,69 +233,81 @@ static struct wlr_box *get_mapping(struct wlr_cursor *cur,
 }
 
 bool wlr_cursor_warp(struct wlr_cursor *cur, struct wlr_input_device *dev,
-		double x, double y) {
+		double lx, double ly) {
 	assert(cur->state->layout);
-	bool result = false;
 
+	bool result = false;
 	struct wlr_box *mapping = get_mapping(cur, dev);
+	if (mapping) {
+		result = wlr_box_contains_point(mapping, lx, ly);
+	} else {
+		result = wlr_output_layout_contains_point(cur->state->layout, NULL,
+			lx, ly);
+	}
+
+	if (result) {
+		cursor_warp_unchecked(cur, lx, ly);
+	}
+
+	return result;
+}
 
+static void cursor_warp_closest(struct wlr_cursor *cur,
+		struct wlr_input_device *dev, double lx, double ly) {
+	struct wlr_box *mapping = get_mapping(cur, dev);
 	if (mapping) {
-		if (wlr_box_contains_point(mapping, x, y)) {
-			cursor_warp_unchecked(cur, x, y);
-			result = true;
+		if (!wlr_box_contains_point(mapping, lx, ly)) {
```

Comment:
1) Is there a reason for this indirection over `closest_{x,y}`?
2) Is this explicit check necessary? Without checking the file, I'd expect this to be the first thing `_contains_point` does

Same in the case below

Reply:
>Is there a reason for this indirection over closest_{x,y}?

`double` vs `int` thing

>Is this explicit check necessary? Without checking the file, I'd expect this to be the first thing _contains_point does

Ditto

Reply:
I'm not quite sure what you are trying to tell me here.
Both `current_{x,y}` and and `l{x,y}` are doubles. What am I missing?

Reply:
`wlr_box_closest_point` takes integers

Reply:
Fixed